File: sungwoopark95/python/GraphBFSDFS/cfds_graph_weighted/weighted_graph.py
from __future__ import annotations
from typing import List
# from prim import MinHeap

class WeightedEdge:
    def __init__(self, end:Vertex, weight:int) -> None:
        self.end = end
        self.weight = weight


class Vertex:
    def __init__(self, name:str) -> None:
        self.name = name
        self.key = 0
        self.parent = None
        self.distance = 0
        self.adjacency_list = []
    
    def GetAdjacencyList(self) -> List[WeightedEdge]:
        return self.adjacency_list
    
    def AddToAdjacentList(self, vertex:Vertex, weight:int) -> None:
        edge = WeightedEdge(end=vertex, weight=weight)
        self.adjacency_list.append(edge)
        
        
class Graph:
    def __init__(self) -> None:
        self.vertices = []
    
    def GenVertex(self, name) -> Vertex:
        v = Vertex(name)
        self.vertices.append(v)
        return v
    
    def GenEdge(self, start:Vertex, end:Vertex, weight:int) -> None:
        start.AddToAdjacentList(vertex=end, weight=weight)
        
    def GetEdge(self) -> List[WeightedEdge]:
        edges = []
        for v in self.vertices:
            for adj in v.GetAdjacencyList():
                edges.append((v, adj.end, adj.weight))
        return edges
    
    def GetVertices(self) -> List[Vertex]:
        return self.vertices
    
    def GetNumVertices(self) -> int:
        return len(self.vertices)
    
    def PrintGraph(self) -> None:
        print("============== Defined Graph ==============")
        for vertex in self.vertices:
            print(f"Vertex {vertex.name}, Adjacency List: ", end='')
            is_first = True
            for adj in vertex.GetAdjacencyList():
                # adj : WeightedEdge object
                if not is_first:
                    print(", ", end='')
                print(f"(Vertex {adj.end.name}, Weight {adj.weight})", end='')
                is_first = False
            print()
    
    def __del__(self) -> None:
        ## unnecessary for python
        for v in self.vertices:
            del v
        del self.vertices

File: sungwoopark95/C++/Graph/weighted_graph/weighted_graph.cpp
#include <iostream>
#include <string>
#include <vector>
#include "weighted_graph.h"
using namespace std;

WeightedEdge::WeightedEdge(Vertex* end, int weight) {
    this->end_ = end;
    this->weight_ = weight;
}

Vertex* WeightedEdge::GetEndVertex() {
    return this->end_; 
}

int WeightedEdge::GetWeight() {
    return this->weight_;
}

Vertex::Vertex(string name) {
    this->name_ = name;
}

string Vertex::GetName() {
    return this->name_;
}

vector<WeightedEdge*> Vertex::GetAdjacencyList() {
    return this->adjacency_list_;
}

void Vertex::AddToAdjacentList(Vertex* vertex, int weight) {
    WeightedEdge* edge = new WeightedEdge(vertex, weight);
    this->adjacency_list_.push_back(edge);
}

Graph::~Graph() {
    // GenVertex에서 동적으로 할당된 각각의 vertex를 해제
    size_t size = this->GetNumVertices();
    for (int i=0;i<size;i++) {
        delete this->vertices_[i];
    }
    // vertices를 비움
    this->vertices_.clear();
}

Vertex* Graph::GenVertex(string name) {
    Vertex* newNode = new Vertex(name);
    this->vertices_.push_back(newNode);
    return newNode;
}

void Graph::GenEdge(Vertex* start, Vertex* end, int weight) {
    start->AddToAdjacentList(end, weight);
}

vector<Vertex*> Graph::GetVertices() {
    return this->vertices_;
}

size_t Graph::GetNumVertices() {
    return this->vertices_.size();
}

void Graph::PrintGraph() {
    cout << "============== Defined Graph ==============" << endl;
    for (auto vertex : vertices_) {
        cout << "Vertex " << vertex->GetName() << ", Adjacency List: ";

        bool is_first = true;
        for (auto adjacent : vertex->GetAdjacencyList()) {
            if (is_first == false) {
                cout << ", ";
            }
            cout << "(Vertex " << adjacent->GetEndVertex()->GetName() << " Weight " << adjacent->GetWeight() << ")";
            is_first = false;
        }
        cout << endl;
    }
}